src/raftstore/store/peer.rs

```diff
@@ -484,6 +494,34 @@ impl Peer {
         StaleState::Valid
     }
 
+    fn next_lease_expired_time(&self, send_to_quorum_ts: Timespec) -> Timespec {
+        // The valid leader lease should be
+        // "lease = election_timeout - (quorum_commit_ts - send_to_quorum_ts)"
+        // And the expired timestamp for that leader lease is "quorum_commit_ts + lease",
+        // which is "send_to_quorum_ts + election_timeout" in short.
+        send_to_quorum_ts + self.election_timeout
+    }
+
+    fn update_leader_lease(&mut self, ready: &Ready) {
+        // Update leader lease variables when the raft state changes
+        if ready.ss.is_some() {
+            let ss = ready.ss.as_ref().unwrap();
+            match ss.raft_state {
+                StateRole::Leader => {
+                    self.leader_lease_expired_time =
```

Comment:
And to be noted, the lease time set here is not the lease produce by election but the time when first empty entry is proposed. No matter how the leader is delay, the first empty entry will not be sent until lease time is updated. So as long as the first empty entry is applied at some point, the lease set here is correct and safe.